core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -188,8 +191,10 @@ class ReplicaFetcherThread(name: String,
         s"leader's start offset $leaderStartOffset")
       val offsetToFetch = Math.max(leaderStartOffset, replica.logEndOffset.messageOffset)
       // Only truncate log when current leader's log start offset is greater than follower's log end offset.
-      if (leaderStartOffset > replica.logEndOffset.messageOffset)
-        replicaMgr.logManager.truncateFullyAndStartAt(topicPartition, leaderStartOffset)
+      if (leaderStartOffset > replica.logEndOffset.messageOffset) {
+        partition.truncateFullyAndStartAt(leaderStartOffset, isFuture = false)
+        replicaMgr.replicaAlterLogDirsManager.markPartitionsForTruncation(brokerConfig.brokerId, Map(topicPartition -> leaderStartOffset))
```

Comment:
Good point. This makes sense. Let me think more about how to handle this case properly.

Reply:
@junrao Sorry for late reply on this issue. Now I think about this more, I think the current patch handles this case well.

I understand that if truncateFullyAndStartAt() is called for the current replica, the current patch will truncate the future replica to high watermark which does not have any effect. But truncateFullyAndStartAt() will also be called for the future replica later. If truncateFullyAndStartAt() is called for the current replica, then we know that futureReplica's logEndOffset < "currentReplica's logEndOffset before truncation" < "currentReplica's logStartOffset after truncation". This means that ReplicaAlterDirThread will see OffsetOutOfRangeException for futureReplica and call handleOffsetOutOfRange for future replica. Later truncateFullyAndStartAt() will be called for the future replica as well, which makes the result correct.

Do you think this make sense? If not, can you explain a bit more what will go wrong?

Reply:
@lindong28 : Good point that we can rely on truncateFullyAndStartAt() being called on the future replica later since its offset will be out of range too. In that case, it seems that  replicaMgr.replicaAlterLogDirsManager.markPartitionsForTruncation() is a no op. Do we need to call it here?

However, the same issue may happen on truncateTo(). Consider the following case. The current replica needs to call partition.truncateTo(leaderEndOffset) since its logEndOffset is > leaderEndOffset, which can happen when an unclean leader election is triggered. We then call replicaMgr.replicaAlterLogDirsManager.markPartitionsForTruncation(). If there is no leader epoch info (e.g., message format is still old), the future replica will default to truncating to its HW, which could be > the leaderEndOffset of the current replica. By the time the future replica fetches again, more data could have been accumulated in the current replica and the future replica's offset could still be in range. However, some of the data in the future replica now are different from the current replica. In this case, it's better if the future replica defaults to the leaderEndOffset of the current replica during truncation.

Reply:
@junrao Yeah you are right, markPartitionsForTruncation() is not needed if currentReplica. truncateFullyAndStartAt() is called.

Regarding the second issue, I agree that the inconsistency between the current and the future replica can happen if the remaining message's format is old. But it seems to me that the same inconsistency can happen between follower and leader as well if the message format is old. Thus the inconsistency between the future and the current replica does not increase the problem we already have. And in the long term, all messages will have the new format and this won't be an issue. Thus I am not sure the reduced change of inconsistency is worth additional code complexity. But I will update the patch as suggested if you think this is worth doing. What do you think?

Reply:
@lindong28 : I agree that the same issue may happen between the leader and the follower, which we can improve separately in the future. We probably don't want to introduce new issues here between the current and the future replica. I am not sure if my suggestion will add more complexity though. We already store the initial offset in AbstractFetcherThread.partitionStates, but ignore it in maybeTruncate(). It seems that it's more natural to fall back to this initial offset instead of HW. Then, we just need to set the initial offset in markPartitionsForTruncation().

Reply:
@junrao The reason I think they are the same problem (not new issue) is that, the future replica is conceptually very similar to the follower replica. ReplicaAlterLogDirsThread works in a very similar way as the ReplicaFetcherThread. They currently share the same problem regarding the inconsistency. And if we have a solution for the follower replica, the solution should be applicable to the future replica as well. I personally think the future replica is just another follower replica on the same broker as leader.

Let's say we think the inconsistency in the future replica is a new problem and we want to fix it. I think re-using the offset in `AbstractFetcherThread.partitionStates` for truncation will make the code a bit more complicated for the following reasons:

- If we allow markPartitionsForTruncation() to change the offset in `partitionStates`, then ReplicaFetcherThread (and ReplicaAlterLogDirsThread) will need to handle this properly in processPartitionData(). Currently only one thread will update the offset in `partitionStates` and `fetchOffset` should always equals `replica.logEndOffset.messageOffset` in `processPartitionData()`.

- As of now, the offset in `partitionStates` is always the LEO of the follower (and future) replica. Thus it is no-op if we truncate the follower replica to this offset in maybeTruncate(). Re-using this offset to store the truncation offset may make the code harder to maintain. It may be better to store the truncationOffsets in a separate map and use it in maybeTruncate(...) as shown in https://github.com/apache/kafka/pull/3874/commits/8c45ede9c81b6509d393941db684e4c7b07585e5#diff-6446cfed30fa23cfc249b9537fffbe35R175.

Thanks for all the discussion. Give my explanation above, can you let me know which solution you prefer? Besides, do you want me to change the name of configs and classes as you raised in the other comments?

Reply:
@lindong28 : A couple of things. First, when markPartitionsForTruncation(), the replica is now transitioning to a maybe truncating state (not ready for fetching). If there is a pending fetch response, we should just ignore it. I have a separate PR that tries to tighten this up. Second, once we have that. The implementation that I was thinking is the following. markPartitionsForTruncation() just marks the partition as maybe truncating and puts the initial offset in partitionStates. From that point on, the replica won't be used for fetching data until it transitions to the ready for fetch state. During the handling of maybeTruncate() logic, we will try to decide the truncation point based on leader epoch if possible. Otherwise, we will just fall back to the initial offset. So, implementation wise, I am not sure we really need a separate truncationOffsets map.

So, my thinking is that such a change seems relatively small and safer than the current approach.

As for the naming, I don't want have a strong preference. It's up to you to pick sth consistent.

Reply:
This solution makes sense. I have updated the patch as suggested.

It seems that all comments have been addressed. I will rebase the patch onto trunk and review it myself tomorrow. Thanks!